### PR TITLE
TBOX-340-add-core-connect-jdbc-ingest

### PR DESCRIPTION
**Pull request by MelissaPang**
# ↪️ Pull Request

## ✔️ PR Todo

- [ ] Add documentation
- [ ] Add examples
- [ ] Ensure you are following the practices in the [contributor guide](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit) and [coding_standards](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards)
  * [Use consistent and descriptive name conventions](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.477s36w4rds9) 
  * [Uses google-style docstrings](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Google-style-docstrings)
  * [Uses versioned API/client methods only](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.yv6df99fyrq2)
  * [Avoid global variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-global-variables)
  * [Avoid giant try/catch](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-giant-try-/-except)
  * [Don't use mutable default variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Don%E2%80%99t-use-mutable-default-arguments)
  * [Use a `main()` function in your scripts](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Use-a-%E2%80%9Cmain()%E2%80%9D--function-in-your-scripts)
- [ ] Added/updated testing for this change (ensure you are testing any changes/new code!)
- [ ] Update title to link to TBOX issue (i.e. start pull request title with TBOX-xx) 
- [ ] Approval from first code-reviewer
- [ ] Approval from second code-reviewer (only required if a second reviewer has left comments)
- [ ] Pass all checks

### CODE REVIEW

File: tamr_toolbox/data_io/core_connect.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        raise Exception(error_message)
+
+    # handle primary key
+    if primary_key is None:
+        primary_key = []
+    else:
+        primary_key = primary_key.split(",")
+
+    # ingest data
+    api_path = "/api/connect/jdbcIngest/ingest"
```
Comment (zbpvarun-tamr) [inline tamr_toolbox/data_io/core_connect.py]: Given that you are starting out with just the jdbc ingest endpoint, I have the following suggestion to you. Why not leverage the existing df-connect code? The code looks almost identical to the df-connect ingest function right? How about adding a version check in the df-connect module. If the version is >=v2022.005, use the core-connect endpoint. Else, use the df-connect endpoint. That way, we do not break functionality. I do envision needing a separate core-connect module as we add other cloud endpoints that are present in core-connect but not df-connect... But the jdbc ingest is common between both. I suggest we treat it as such.

File: tests/data_io/test_core_connect.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        assert not tbox.dataset.manage.exists(client=client, dataset_name=dataset_name)
+
+
+@mock_api(enforce_online_test=enforce_online_test)
+def test_ingest_dataset():
+    tamr_client = utils.client.create(**CONFIG["toolbox_test_instance"])
+    jdbc_connect = CONFIG["core_connect"]["jdbc"]["ingest"]
+
+    # Reset test datasets if they exist
+    # remove_test_datasets(client=tamr_client)
```
Comment (zbpvarun-tamr) [inline tests/data_io/test_core_connect.py]: Why is this line commented out?

File: tests/data_io/test_core_connect.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    tbox.data_io.core_connect.ingest_dataset(
+        client=tamr_client,
+        jdbc_connect=jdbc_connect,
+        dataset_name=DATASET_NAME,
+        query="select * from dataset.dataset_ns_current limit 10",
+        primary_key=PRIMARY_KEYS,
+        # truncate_tamr_dataset="true",
+        retrieve_connect_metadata="true",
+    )
```
Comment (zbpvarun-tamr) [inline tests/data_io/test_core_connect.py:42]: You will want to do a little more here - an assert statement that the dataset now exists in Tamr for example. And potentially rerun the remove_test_datasets function.

File: tests/data_io/test_core_connect.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+CONFIG = utils.config.from_yaml(
+    get_toolbox_root_dir() / "tests/mocking/resources/toolbox_test.yaml"
+)
+
+DATASET_NAME = "test_core_connect"
+PRIMARY_KEYS = None
+
+enforce_online_test = False
+
+
```
Comment (zbpvarun-tamr) [inline tests/data_io/test_core_connect.py:17]: You should probably add some tests to ensure that the various error handling you have built in the ingest function actually trigger.

File: tests/mocking/resources/toolbox_test.yaml [hunk 5]
```diff
@@ -55,4 +55,17 @@ df_connect:
...
+    protocol: "http"
+    port: "9050"
+    tamr_username: "admin"
+    tamr_password: $TAMR_TOOLBOX_PASSWORD
+    jdbc:
+        ingest:
+            jdbcUrl: $CONNECT_TOOLBOX_POSTGRES_JDBC
+            dbUsername: "tamr"
+            dbPassword: $CONNECT_TOOLBOX_DB_PASSWORD
+            fetchSize: 10000
```
Comment (zbpvarun-tamr) [inline tests/mocking/resources/toolbox_test.yaml]: You're missing documentation. But I suppose that can wait until the other comments are addressed.

**Review by zbpvarun-tamr — CHANGES_REQUESTED** [review]

File: tests/data_io/test_core_connect.py (hunk 2 above)
Comment (MelissaPang) [reply tests/data_io/test_core_connect.py]: Forgot to uncomment it. Fixed.

File: tests/data_io/test_core_connect.py (hunk 3 above)
Comment (MelissaPang) [reply tests/data_io/test_core_connect.py:42]: Assert statements are added: one assert statement for testing if an imported dataset exists; one assert statement for importing a dataset with pk and truncating the tamr dataset;

File: tests/data_io/test_core_connect.py (hunk 4 above)
Comment (MelissaPang) [reply tests/data_io/test_core_connect.py:17]: Tests are added. One test for raising the error of Tamr version; another test for raising the error when an ingestion fails.

File: tests/mocking/resources/toolbox_test.yaml (hunk 5 above)
Comment (MelissaPang) [reply tests/mocking/resources/toolbox_test.yaml]: Will add documentation when other comments are addressed.

File: tamr_toolbox/data_io/core_connect.py (hunk 1 above)
Comment (MelissaPang) [reply tamr_toolbox/data_io/core_connect.py]: I agree that there will be more functions added in the core-connect module for other ingestion and exporting functions. I would prefer to keep the core-connect ingest function in core-connect module for a couple reasons: 1.  it is confusing to call a core-connect ingestion function from the df-connect module; 2. there are more options for core-connect ingestion function than df-connect; 3. it is easy to test the functions separately.

Comment (zbpvarun-tamr) [reply tamr_toolbox/data_io/core_connect.py]: Fair points. If so, then I would still suggest that you do the following:

1. Rename the ingest_dataset function to jdbc_ingest or something along the lines so it's very clear which ingest function among all the endpoints in core-connect that this is referring to.
2. Add an optional flag argument (default true) that if true, will redirect to the df-connect endpoint if the version is before 2022.005. If false, let it fail with the error message you have plus info about the flag argument.

File: tamr_toolbox/data_io/core_connect.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    dataset_name: str,
+    query: str,
+    primary_key=None,
+    truncate_tamr_dataset=None,
+    retrieve_connect_metadata=None,
+    retrieve_source_metadata=None,
+    tamr_min_version="2022.005.0",
+) -> JsonDict:
+    """
+    Ingest a dataset into Tamr via core_connect given query config, dataset name, query string,
```
Comment (zbpvarun-tamr) [inline tamr_toolbox/data_io/core_connect.py]: Specify that this function uses the jdbc ingest endpoint. Assume that there will be other functions in this module for the other ingest endpoints.

File: tests/mocking/resources/toolbox_test.yaml (hunk 5 above)
Comment (zbpvarun-tamr) [reply tests/mocking/resources/toolbox_test.yaml]: You should add documentation in your next push

**Review by zbpvarun-tamr — CHANGES_REQUESTED** [review]

File: tamr_toolbox/data_io/core_connect.py (hunk 1 above)
Comment (MelissaPang) [reply tamr_toolbox/data_io/core_connect.py]: 1. function renamed to jdbc_ingest.
2. As we cannot test df-connect endpoint and core-connect endpoint in the same instance, we will leave df-connect and core-connect separate modules.

File: tests/mocking/resources/toolbox_test.yaml (hunk 5 above)
Comment (MelissaPang) [reply tests/mocking/resources/toolbox_test.yaml]: documentation for jdbc_ingest is added.

File: tamr_toolbox/data_io/core_connect.py (hunk 1 above)
Comment (zbpvarun-tamr) [reply tamr_toolbox/data_io/core_connect.py]: Resolved internally - will not maintain backwards compatibility with df-connect.

File: doc_src/examples/data_io.rst [hunk 7]
```diff
@@ -57,6 +57,22 @@ Check if Tamr dataset is streamable
 .. literalinclude:: ../../examples/snippets/data_io/df_connect/streamable_check.py
   :language: python
 
+
+Core-Connect
+------------------------------------------------------------------------
+Create a Tamr client and ingest a table as a Tamr dataset
```
Comment (zbpvarun-tamr) [inline doc_src/examples/data_io.rst]: Make it clear that this example is for JDBC ingest. There are a lot of other ingest endpoints in core-connect.

File: tamr_toolbox/data_io/core_connect.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        raise Exception(error_message)
+
+    # handle primary key
+    if primary_key is None:
+        primary_key = []
+    else:
+        primary_key = primary_key.split(",")
+
+    # ingest data
+    api_path = "/api/connect/jdbcIngest/ingest"
```
Comment (zbpvarun-tamr) [inline tamr_toolbox/data_io/core_connect.py]: Add the async=true argument to the URL and include job monitoring using the jobs APIs. Feel free to copy my existing function: https://github.com/Datatamer/ThermoFisher/blob/master/airflow_plugins/unify_plugin/playbooks/thermo_fisher_SpendHQ_import.py#L67

**Review by zbpvarun-tamr — CHANGES_REQUESTED** [review]
So I think we're close to done. Just would like some quality of life improvements for others who try to add to this module in the future. As discussed over Slack, please do the following restructuring:

1. Move core-connect to its own folder under data_io, just as df-connect is.
2. Create a secondary script for common functions and you can put the job monitoring function in there.
3. Rename your primary script and function (if needed) so as to make sense when calling it. Maybe call the script ingest.py and the function jdbc_ingest?

**Review by zbpvarun-tamr — CHANGES_REQUESTED** [review]
One last thing - can you move the "run_and_poll_connect_job" function to a new file like common.py or something? It will be used for every core-connect endpoint and so, when we have multiple files corresponding to the various services, we would need to import this function into all of them.

**Comment by MelissaPang** [comment]
> One last thing - can you move the "run_and_poll_connect_job" function to a new file like common.py or something? It will be used for every core-connect endpoint and so, when we have multiple files corresponding to the various services, we would need to import this function into all of them.

Yes, I moved the "_run_and_poll_connect_job" function to the /tamr_toolbox/common.py.

File: tamr_toolbox/data_io/core_connect.py (hunk 8 above)
Comment (MelissaPang) [reply tamr_toolbox/data_io/core_connect.py]: Yes, async=true is added as below
api_path = "/api/connect/jdbcIngest/ingest?async=true"

File: doc_src/examples/data_io.rst (hunk 7 above)
Comment (MelissaPang) [reply doc_src/examples/data_io.rst]: Yes, line 63 is updated to reflect the example is for JDBC ingest. "Create a Tamr client and ingest a table as a Tamr dataset via JDBC"

**Comment by MelissaPang** [comment]
> So I think we're close to done. Just would like some quality of life improvements for others who try to add to this module in the future. As discussed over Slack, please do the following restructuring:
> 
> 1. Move core-connect to its own folder under data_io, just as df-connect is.
> 2. Create a secondary script for common functions and you can put the job monitoring function in there.
> 3. Rename your primary script and function (if needed) so as to make sense when calling it. Maybe call the script ingest.py and the function jdbc_ingest?

1. Yes, core-connect is moved to its own folder and renamed. /tamr_toolbox/data_io/core-connect/jdbc.py. All jdbc ingestion and export will be written in the jdbc.py script. There will be separate script for other core-connect services, such as s3.py, gcs.py, and etc.
2. common function (e.g. _run_and_poll_connect_job) is added to the /tamr_toolbox/common.py.
3. Primary script is renamed to jdbc.py. The jdbc ingestion function is renamed to jdbc_ingest.

File: tamr_toolbox/data_io/core_connect/jdbc.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        HTTPError: if the call to ingest the dataset was unsuccessful
+    """
+
+    # Check Tamr version
+    if version.is_version_condition_met(
+        tamr_version=version.current(client), min_version=tamr_min_version
+    ):
+        None
+    else:
+        error_message = "Core-connect is not available in current version of Tamr."
```
Comment (zbpvarun-tamr) [inline tamr_toolbox/data_io/core_connect/jdbc.py]: Being really picky here but can you change this to say "Toolbox does not support core-connect for current version of Tamr"? Don't want people on v2022.005 - v2022.010 to come back and say something.

Comment (MelissaPang) [reply tamr_toolbox/data_io/core_connect/jdbc.py]: Sure, it is updated.

**Review by zbpvarun-tamr — DISMISSED** [review]
Awesome work!

**Review by zbpvarun-tamr — APPROVED** [review]

**Pull request merged**